src/blueapi/service/authentication.py

```diff
@@ -0,0 +1,264 @@
+import base64
+import json
+import os
+import time
+from enum import Enum
+from http import HTTPStatus
+from typing import Any
+
+import jwt
+import requests
+from dotenv import load_dotenv
+
+
+class AuthenticationType(Enum):
+    DEVICE = "device"
+    PKCE = "pkce"
+
+
+class Authentication:
+    """
+    Authentication class handles the authentication process using either
+    device code flow or PKCE flow.
+
+    Attributes:
+        client_id (str): The client ID for the authentication.
+        authentication_url (str): The URL for authentication.
+        audience (list[str]): The audience for the authentication.
+        token_url (str): The URL to obtain the token.
+        openid_config (str): The OpenID configuration URL.
+        issuer (str): The issuer of the token.
+        user_name (str | None): The username obtained from the token.
+        fedid (str | None): The federated ID obtained from the token.
+        jwks_client (jwt.PyJWKClient): The JWKS client for verifying tokens.
+        token_file_path (str): The file path to save the token.
+        token (None | dict[str, Any]): The token dictionary.
```

Comment:
As commented elsewhere, split into two: client-side authentication request and server-side authentication checking.

Which of these fields are actually required for each?
Should get the value of the fields from the ApplicationConfig, which then means they can be overriden with standard handling.